compiler/noirc_frontend/src/ownership/last_uses.rs

```diff
@@ -0,0 +1,410 @@
+            Expression::Loop(loop_expr) => self.track_variables_in_loop(loop_expr),
+            Expression::While(while_expr) => self.track_variables_in_while(while_expr),
+            Expression::If(if_expr) => self.track_variables_in_if(if_expr),
+            Expression::Match(match_expr) => self.track_variables_in_match(match_expr),
+            Expression::Tuple(elems) => self.track_variables_in_tuple(elems),
+            Expression::ExtractTupleField(tuple, _index) => {
+                self.track_variables_in_expression(tuple);
+            }
+            Expression::Call(call) => self.track_variables_in_call(call),
+            Expression::Let(let_expr) => self.track_variables_in_let(let_expr),
+            Expression::Constrain(boolean, _location, msg) => {
+                self.track_variables_in_constrain(boolean, msg);
+            }
+            Expression::Assign(assign) => self.track_variables_in_assign(assign),
+            Expression::Semi(expr) => self.track_variables_in_expression(expr),
+            Expression::Clone(_) => unreachable!("last_uses is called before clones are inserted"),
+            Expression::Drop(_) => unreachable!("last_uses is called before drops are inserted"),
+            Expression::Break => (),
+            Expression::Continue => (),
+        }
+    }
+
+    /// Under the experimental alternate ownership scheme, whenever an ident is used it is
+    /// always cloned unless it is the last use of the ident (not in a loop). To simplify this
+    /// analysis we always clone here then remove the last clone later if possible.
+    fn track_variables_in_ident(&mut self, ident: &ast::Ident) {
+        // We only track last uses for local variables, globals are always cloned
+        if let ast::Definition::Local(local_id) = &ident.definition {
+            self.remember_use_of_variable(*local_id, ident.id);
+        }
+    }
+
+    fn track_variables_in_literal(&mut self, literal: &Literal) {
+        match literal {
+            Literal::Integer(..) | Literal::Bool(_) | Literal::Unit | Literal::Str(_) => (),
+
+            Literal::FmtStr(_, _, captures) => self.track_variables_in_expression(captures),
+
+            Literal::Array(array) | Literal::Slice(array) => {
+                for element in array.contents.iter() {
+                    self.track_variables_in_expression(element);
+                }
+            }
+        }
+    }
+
+    fn track_variables_in_unary(&mut self, unary: &ast::Unary) {
+        self.track_variables_in_expression(&unary.rhs);
+    }
+
+    fn track_variables_in_binary(&mut self, binary: &ast::Binary) {
+        self.track_variables_in_expression(&binary.lhs);
+        self.track_variables_in_expression(&binary.rhs);
+    }
+
+    /// - Extracting an array from another array (`let inner: [_; _] = array[0];`):
+    ///   - Extracting a nested array from its outer array will always increment the reference count
+    ///     of the nested array.
```

Comment:
Is this comment relevant?

Reply:
No - removed. It was originally pasted from the ownership pass